cpp/src/arrow/dataset/partition.h

```diff
@@ -0,0 +1,191 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/dataset/dataset.h"
+#include "arrow/dataset/type_fwd.h"
+#include "arrow/dataset/visibility.h"
+
+namespace arrow {
+namespace dataset {
+
+// ----------------------------------------------------------------------
+// Computing partition values
+
+// TODO(wesm): API for computing partition keys derived from raw
+// values. For example, year(value) or hash_function(value) instead of
+// simply value, so a dataset with a timestamp column might group all
+// data with year 2009 in the same partition
+
+// /// \brief
+// class ScalarTransform {
+//  public:
+//   virtual Status Transform(const std::shared_ptr<Scalar>& input,
+//                            std::shared_ptr<Scalar>* output) const = 0;
+// };
+
+// class PartitionField {
+//  public:
+
+//  private:
+//   std::string field_name_;
+// };
+
+// ----------------------------------------------------------------------
+// Partition identifiers
+
+/// \brief A partition level identifier which can be used
+///
+/// TODO(wesm): Is this general enough? What other kinds of partition
+/// keys exist and do we need to support them?
+class PartitionKey {
+ public:
+  const std::vector<std::string>& fields() const { return fields_; }
+  const std::vector<std::shared_ptr<Scalar>>& values() const { return values_; }
```

Comment:
I'm not sure what PartitionKey implements. Just `AND(field==value for field, value in zip(fields, values))`?

Reply:
Yes, that's probably just it. This part is not complete, I presume that generalization will be required as time goes on. In the simple case partitions have a single field and value (that's for Hive-style partitioning, but we already know that this is not general enough)